src/msg.rs

```diff
@@ -256,6 +257,78 @@ pub enum ExecuteMsg {
     },
 }
 
+pub trait Decoyable {
+    fn get_minimal_decoys_size(&self) -> usize;
+    fn get_entropy(self) -> Option<Binary>;
+}
+
+impl Decoyable for ExecuteMsg {
+    fn get_minimal_decoys_size(&self) -> usize {
+        match self {
+            ExecuteMsg::Deposit { decoys, .. }
+            | ExecuteMsg::Redeem { decoys, .. }
+            | ExecuteMsg::Transfer { decoys, .. }
+            | ExecuteMsg::Send { decoys, .. }
+            | ExecuteMsg::Burn { decoys, .. }
+            | ExecuteMsg::Mint { decoys, .. }
+            | ExecuteMsg::TransferFrom { decoys, .. }
+            | ExecuteMsg::SendFrom { decoys, .. }
+            | ExecuteMsg::BurnFrom { decoys, .. } => {
+                if let Some(user_decoys) = decoys {
+                    return user_decoys.len();
+                }
+
+                0
+            }
+            ExecuteMsg::BatchSendFrom { actions, .. } => get_min_decoys_count(actions),
+            ExecuteMsg::BatchTransferFrom { actions, .. } => get_min_decoys_count(actions),
+            ExecuteMsg::BatchTransfer { actions, .. } => get_min_decoys_count(actions),
+            ExecuteMsg::BatchSend { actions, .. } => get_min_decoys_count(actions),
+            ExecuteMsg::BatchBurnFrom { actions, .. } => get_min_decoys_count(actions),
+            ExecuteMsg::BatchMint { actions, .. } => get_min_decoys_count(actions),
+            _ => 0,
+        }
+    }
+
+    fn get_entropy(self) -> Option<Binary> {
+        match self {
+            ExecuteMsg::Deposit { entropy, .. }
+            | ExecuteMsg::Redeem { entropy, .. }
+            | ExecuteMsg::Transfer { entropy, .. }
+            | ExecuteMsg::Send { entropy, .. }
+            | ExecuteMsg::Burn { entropy, .. }
+            | ExecuteMsg::Mint { entropy, .. }
+            | ExecuteMsg::TransferFrom { entropy, .. }
+            | ExecuteMsg::SendFrom { entropy, .. }
+            | ExecuteMsg::BurnFrom { entropy, .. }
+            | ExecuteMsg::BatchTransferFrom { entropy, .. }
+            | ExecuteMsg::BatchSendFrom { entropy, .. }
+            | ExecuteMsg::BatchTransfer { entropy, .. }
+            | ExecuteMsg::BatchSend { entropy, .. }
+            | ExecuteMsg::BatchBurnFrom { entropy, .. }
+            | ExecuteMsg::BatchMint { entropy, .. } => entropy,
+            _ => None,
+        }
+    }
+}
+
+fn get_min_decoys_count<T: HasDecoy>(actions: &[T]) -> usize {
+    let mut min_decoys_count = usize::MAX;
+    for action in actions {
+        if let Some(user_decoys) = &action.decoys() {
+            if user_decoys.len() < min_decoys_count {
+                min_decoys_count = user_decoys.len();
+            }
+        }
+    }
+
+    if min_decoys_count == usize::MAX {
```

Comment:
maybe a bit more clear is to check here if actions.len() == 0.
also has the advantage that if the user somehow sent usize::MAX decoys, it doesn't convert to 0